### DISCUSSION TITLE
A bit unusual syntax

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#154] [General]
```ruby
class HelloWorld {
  function : Main(args : String[]) ~ Nil {
    hellos := ["Hello World", "Καλημέρα κόσμε", "こんにちは 世界"];
    each(i : hellos) {
      hellos[i]->PrintLine();
    };
  }
}
```

`each(i : hellos)` should be understand as for each i in the array hellos (i is an element of the array hellos), so i is definitely not the index.

On other languages, it will look like this:

```ruby
class HelloWorld {
  function : Main(args : String[]) ~ Nil {
    hellos := ["Hello World", "Καλημέρα κόσμε", "こんにちは 世界"];
    each(i : hellos) {
      i->PrintLine();
    };
  }
}
```

**Comment by objeck** [comment]
Here's the [code](https://github.com/objeck/objeck-lang/blob/master/core/compiler/parser.cpp#L4658) to change it.

**Comment by objeck** [comment]
Here would be the idea.

Update the value of ``count_ident`` to something like ``"#" + count_ident + "_index"`` so it is unique within nested loops. Decorate the ``For* for_stmt`` tree node and [carry over](https://github.com/objeck/objeck-lang/blob/master/core/compiler/context.cpp#L3926) the original variable name for ``count_ident``. Use the presence of the ``count_ident`` value as a flag to add a new assignment statement. The new statement must determine if you are dealing with an array or object instance with a ``Get(..)`` method, so the parser cannot do it.

Last, consider a new syntax such as ``each(hello := hellos) { ... };`` as the original syntax is in use. Estimate a weekend effort with regression testing to ensure that nothing breaks.

**Reply by ghost** [reply]
I will not propose a new syntax to cause more confusions. But if I wanted to propose one, I will propose `foreach`:

```ruby
class HelloWorld {
  function : Main(args : String[]) ~ Nil {
    hellos := ["Hello World", "Καλημέρα κόσμε", "こんにちは 世界"];
    foreach(i : hellos) {
      i->PrintLine();
    };
  }
}
```

I will propose `rforeach` for `reverse` equivalent if you ask me. But as I said, I will not propose a new syntax to cause more confusions.

**Comment by ghost** [comment]
The question is why it's this way from the beginning. You have never used any functional programming languages, don't you? You tried to copy the look and feel but you failed to understand the way it really works.

A more functional style for the above code:

`hellos.each |String i| { i->PrintLine(); }`

**Comment by objeck** [comment]
The ``each`` syntax has nothing to do with functional programming. 

Instead, it had to do with making ``for`` loops more straightforward. Also, finding a path to make _nearly_ all code generated map to machine code, thus the tradeoffs. 

Thank you for the feedback.

**Comment by objeck** [comment]
Feature added in v2023.2.2
